resources/js/features/game-list/components/GameListDataTable/GameListDataTable.tsx

```diff
@@ -1,87 +1,172 @@
-import type { Table } from '@tanstack/react-table';
-import { flexRender } from '@tanstack/react-table';
+import type { Row, Table } from '@tanstack/react-table';
+import { useEffect, useRef } from 'react';
 import { useTranslation } from 'react-i18next';
 
 import {
   BaseTable,
   BaseTableBody,
   BaseTableCell,
-  BaseTableHead,
-  BaseTableHeader,
   BaseTableRow,
 } from '@/common/components/+vendor/BaseTable';
 import { cn } from '@/common/utils/cn';
 
-interface GameListDataTableProps<TData> {
-  table: Table<TData>;
+import { GameRow } from './GameRow';
+import { SystemHeaderRow } from './SystemHeaderRow';
+import { TableHeader } from './TableHeader';
+
+interface GameListDataTableProps {
+  table: Table<App.Platform.Data.GameListEntry>;
+
+  /** Affects row grouping behavior and whether or not a loading state is displayed. */
+  isLoading?: boolean;
 }
 
-export function GameListDataTable<TData>({ table }: GameListDataTableProps<TData>) {
+export function GameListDataTable({ table, isLoading = false }: GameListDataTableProps) {
   const { t } = useTranslation();
 
+  const wasShowingGroups = useRef(false);
+  const lastSortState = useRef(false);
+
   const visibleColumnCount = table.getVisibleFlatColumns().length;
+  const rows = table.getRowModel().rows;
+
+  const isSortedBySystem = table.getState().sorting.some((sort) => sort.id === 'system');
+
+  useEffect(() => {
+    lastSortState.current = isSortedBySystem;
+  }, [isSortedBySystem]);
+
+  /**
+   * Determines whether games should be grouped by system based on the current table state.
+   * Groups are shown when all of the following are true:
+   * 1. The table is sorted by system.
+   * 2. Multiple systems are present on the current page.
+   * 3. The table is not in a loading state.
+   */
+  const getShouldGroupBySystem = (): boolean => {
+    if (isLoading && wasShowingGroups.current) {
+      return true;
+    }
+
+    if (isLoading && isSortedBySystem && !lastSortState.current) {
+      return false;
+    }
+
+    const hasMultipleSystems = rows?.length > 0 && new Set(rows.map(getSystemName)).size > 1;
```

Comment:
This logic only applies to the data on the current page. If I set my "per page" to 10 and start paging through it, the system header disappears on pages that only match a single system.

Additionally, if the data rolls over to the next page, the counter only shows the number for the current page (for example, there's 3 more on the next page):
![image](https://github.com/user-attachments/assets/a97c50dc-682f-44ea-8ae1-8f0e86cca789)

I imagine that using true counts regardless of the pagination wouldn't be worth the effort, but I don't imagine this sort will get used much outside of the Developed Sets page (which hasn't been converted yet).

Reply:
Unfortunately, trying to make this work across pages slows the DB query down by 20x regardless of what SQL incantations I try.

Reply:
In latest, I've removed the game count from the header rows.

Reply:
It's still weird that it disappears on pages that only display games for a single system, but I'm not sure of a good solution for paging with headers.